code/client/munkilib/middleware.py

```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/python
+
+import os
+import subprocess
+from threading import Timer
+
+import munkicommon
+import fetch
+
+
+"""
+To use this module you must provide a path to the middleware
+you wish to execute. You must also be able to provide responses
+for these 2 arguments.
+
+/path/to/middleware [arg1] [arg2]
+
+arg1: the URL
+arg2: is either 'headers' or 'url'
+    When 'headers' is called:
+        The middleware must return one header per line to stdout
+    When 'url' is called:
+        The middleware must return the modified URL to stdout.
+
+"""
+
+MIDDLEWARE_PATH = munkicommon.pref('Middleware')
+
+
+class Exec(Exception):
+    """General exception for middleware"""
+    pass
+
+class Timeout(Exception):
+    """Timeout exception for middleware"""
+    pass
+
+def kill_proc(proc, timeout):
+    timeout["value"] = True
+    proc.kill()
+
+def cmd_with_timeout(cmd, timeout_sec):
```

Comment:
munkicommon already has a subclass of Popen that supports timeouts; can you use that?
https://github.com/munki/munki/blob/master/code/client/munkilib/munkicommon.py#L150

(I'm not actually convinced you need to have a timeout for this, but if you are, it would be nice to use existing/tested code)